### PR TITLE
notice when App Engine Java components aren't installed

### PR DESCRIPTION
**Pull request by elharo**
@briandealwis

### CODE REVIEW

File: plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/SdkUiMessages.java [hunk 1]
```diff
@@ -4,6 +4,9 @@
 import org.eclipse.osgi.util.NLS;
 
 public class SdkUiMessages extends NLS {
+  
+  // todo Ick! Mutable public fields. Use the other kind of messages.
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/SdkUiMessages.java]: The fields are set from the messages

File: plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/messages.properties [hunk 2]
```diff
@@ -5,6 +5,7 @@ CloudSdkPreferencePage_3=URL malformed
 CloudSdkPreferencePage_4=Cannot launch a browser
 CloudSdkPreferencePage_5=&SDK location:
 CloudSdkPreferencePage_6=No SDK found: {0}
+AppEngineJavaComponentsNotInstalled=App Engine Java components not installed
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/messages.properties]: We should give some help on how to add the java components ('_please run `gcloud install app-engine-java`_ from the command line_'?)

File: plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/CloudSdkPreferenceArea.java [hunk 3]
```diff
@@ -148,11 +149,14 @@ private Path getDefaultSdkLocation() {
...
 
-  protected boolean validateSdk(Path location) {
+  private boolean validateSdk(Path location) {
     try {
       CloudSdk sdk = new CloudSdk.Builder().sdkPath(location).build();
       sdk.validateCloudSdk();
       sdk.validateAppEngineJavaComponents(); 
+    } catch (AppEngineJavaComponentsNotInstalledException ex) {
+      status = new Status(IStatus.WARNING, getClass().getName(),
+          MessageFormat.format(SdkUiMessages.AppEngineJavaComponentsNotInstalled, ex.getMessage()));
```
Comment (briandealwis) [inline plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/CloudSdkPreferenceArea.java]: An alternative here might be to have a hidden `Label` that we expose in this case that contains instructions to do an update.

Or perhaps have a `Link` to trigger the update on their behalf?

> Your Cloud SDK installation is missing required App Engine Java components.  &lt;a&gt;Click to install&lt;/a&gt;.

File: plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/SdkUiMessages.java (hunk 1 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/SdkUiMessages.java]: Yes, and they can be set from other places too. These are all public and mutable.

File: plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/CloudSdkPreferenceArea.java (hunk 3 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/CloudSdkPreferenceArea.java]: I don't think we could trigger an update from a link.

File: plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/messages.properties (hunk 2 above)
Comment (elharo) [inline plugins/com.google.cloud.tools.eclipse.sdk.ui/src/com/google/cloud/tools/eclipse/sdk/ui/preferences/messages.properties]: Not enough space here for that message. :-( Maybe with the hidden label?

**Comment by elharo** [comment]
I'm going to commit this, but we should improve the messaging. I'll add that to the bug.

**Pull request merged**